nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2464 @@
+//!   instance record if the instance does *not* have an active VMM.
+//! - The presence of a migration ID prevents an `instance_migrate` saga from
+//!   succeeding until the current migration is resolved (either completes or
+//!   fails).
+//! - Only the `instance_start` saga can set the instance's *active* Propolis
+//!   ID, and it can only do this if there is currently no active Propolis.
+//! - Only the `instance_migrate` saga can set the instance's *target* Propolis
+//!   ID and migration ID, and it can only do that if these fields are unset.
+//! - Only the `instance_update` saga can unset a migration ID and target
+//!   Propolis ID, which it will do when handling an update from sled-agent that
+//!   indicates that a migration has succeeded or failed.
+//! - Only the `instance_update` saga can unset an instance's active Propolis
+//!   ID, which it will do when handling an update from sled-agent that
+//!   indicates that the VMM has been destroyed (peacefully or violently).
+//!
+//! For the most part, this state machine prevents race conditions where
+//! multiple sagas mutate the same fields in the instance record, because the
+//! states from which a particular transition may start limited. However, this
+//! is not the case for the `instance-update` saga, which may need to run any
+//! time a sled-agent publishes a new instance state. Therefore, this saga has
+//! the dubious honor of using the only distributed lock in Nexus (at the time
+//! of writing), the "instance updater lock".
+//!
+//! ### The Instance-Updater Lock, or, "Distributed RAII"
+//!
+//! Distributed locks [are scary][dist-locking]. One of the *scariest* things
+//! about distributed locks is that a process can die[^1] while holding a lock,
+//! which results in the protected resource (in this case, the `instance`
+//! record) being locked forever.[^2] It would be good for that to not happen.
+//! Fortunately, *if* (and only if) we promise to *only* ever acquire the
+//! instance-updater lock inside of a saga, we can guarantee forward progress:
+//! should a saga fail while holding the lock, it will unwind into a reverse
+//! action that releases the lock. This is essentially the distributed
+//! equivalent to holding a RAII guard in a Rust program: if the thread holding
+//! the lock panics, it unwinds its stack, drops the [`std::sync::MutexGuard`],
+//! and the rest of the system is not left in a deadlocked state. As long as we
+//! ensure that the instance-updater lock is only ever acquired by sagas, and
+//! that any saga holding a lock will reliably release it when it unwinds, we're
+//! ... *probably* ... okay.
+//!
+//! When an `instance-update` saga is started, it attempts to [acquire the
+//! updater lock][instance_updater_lock]. If the lock is already held by another
+//! update saga, then the update saga completes immediately. Otherwise, the saga
+//! then queries CRDB for a snapshot of the current state of the `instance``
```

Comment:
Yeah, I was using it in the sense of "it represents the current state of the instance and associated records at the time we acquired the lock", not the sense of a ZFS snapshot.

Unfortunately, we already have a type called `InstanceAndVmm`, which is just an instance and its active VMM. This type, on the other hand, contains an instance, its active VMM, its migration target VMM (if there is one), and its currently active migration (if there is one). We *could* just make all the existing code that currently uses just `InstanceAndVmm` use the "instance and active VMM and target VMM and migration" type, but it seems not great to do additional JOINs every time we just want an instance and its active VMM. So, the alternative is to come up with _some_ name for this new thing...

I would really prefer not to call it `InstanceAndActiveVmmAndTargetVmmAndMigration`, because that seems excessive even for the present Nexus naming idioms. :) I had considered `InstanceGestalt` --- it's the totality of db records that represent the entire current state of the instance --- but decided that was too goofy, and settled on "snapshot"...but I'm very open to suggestions here!

Reply:
I suppose it's not the end of the world to call this a snapshot, it just struck me as funny -- all the "views" of the database are snapshots in some sense, so it was a little quirky to have this one be given a distinct name.

... That being said, I do acknowledge, some of these alternative names are worse, so maybe I just need to re-acclimate to the name :sweat_smile:

Reply:
(also to get it officially on the record - I don't mind `InstanceGestalt`! It's a little silly, but it definitely won't be confused with snapshots)